src/lib.rs

```diff
@@ -9,11 +9,145 @@
+                let sess = &state.session;
+                let mut ls = sess.lint_store.borrow_mut();
+                for pass in early_lint_passes {
+                    ls.register_early_pass(Some(sess), true, pass);
+                }
+                for pass in late_lint_passes {
+                    ls.register_late_pass(Some(sess), true, pass);
+                }
+
+                for (name, to) in lint_groups {
+                    ls.register_group(Some(sess), true, name, to);
+                }
+
+                sess.plugin_llvm_passes.borrow_mut().extend(llvm_passes);
+                sess.mir_passes.borrow_mut().extend(mir_passes);
+                sess.plugin_attributes.borrow_mut().extend(attributes);
+            }
+            old(state);
+        });
+
+        control
+    }
+}
+
+use std::path::Path;
+
+pub fn main() {
+    use std::env;
+
+    if env::var("CLIPPY_DOGFOOD").map(|_| true).unwrap_or(false) {
+        return;
+    }
+
+    let dep_path = env::current_dir().expect("current dir is not readable").join("target").join("debug").join("deps");
+    let sys_root = match (option_env!("MULTIRUST_HOME"), option_env!("MULTIRUST_TOOLCHAIN")) {
+        (Some(home), Some(toolchain)) => format!("{}/toolchains/{}", home, toolchain),
+        _ => option_env!("SYSROOT").expect("need to specify SYSROOT env var during clippy compilation or use multirust").to_owned(),
+    };
+
+    if let Some("clippy") = std::env::args().nth(1).as_ref().map(AsRef::as_ref) {
+        let args = wrap_args(std::env::args().skip(2), dep_path, sys_root);
+        let path = std::env::current_exe().expect("current executable path invalid");
+        let run = std::process::Command::new("cargo")
+            .args(&args)
+            .env("RUSTC", path)
+            .spawn().expect("could not run cargo")
+            .wait().expect("failed to wait for cargo?")
+            .success();
+        assert!(run, "cargo rustc failed");
+    } else {
+        let args: Vec<String> = if env::args().any(|s| s == "--sysroot") {
+            env::args().collect()
+        } else {
+            env::args().chain(Some("--sysroot".to_owned())).chain(Some(sys_root)).collect()
+        };
+        rustc_driver::run_compiler(&args, &mut ClippyCompilerCalls::new());
+    }
+}
+
+fn wrap_args<P, I>(old_args: I, dep_path: P, sysroot: String) -> Vec<String>
+    where P: AsRef<Path>, I: Iterator<Item=String> {
+
+    let mut args = vec!["rustc".to_owned()];
+
+    let mut found_dashes = false;
+    for arg in old_args {
+        found_dashes |= arg == "--";
+        args.push(arg);
+    }
+    if !found_dashes {
+        args.push("--".to_owned());
+    }
+    args.push("-L".to_owned());
+    args.push(dep_path.as_ref().to_string_lossy().into_owned());
+    args.push(String::from("--sysroot"));
+    args.push(sysroot);
+    args.push("-Zno-trans".to_owned());
+    args
 }
```

Comment:
I noticed the comment, there was some discussion (that disappeared) and some discussion on irc, it's really hard to do this without splitting the main clippy stuff into its own crate, because I couldn't figure out how to let `src/main.rs` link against `clippy` statically, because clippy sets `plugin = true` and then that's that, no way to also compile it as a `staticlib`. So we'd need to split clippy into a crate with all the code, and a plugin crate + binary crate (the last two can be together again)... or, I just noticed, the plugin crate can link against the binary crate, but that's probably even messier than what we have now.